docs/source/kedro_project_setup/code_beyond_starter_files.md

```diff
@@ -0,0 +1,133 @@
+# Adding code beyond starter files
+
+After you [create a Kedro project](../get_started/new_project.md) and
+[add a pipeline](../tutorial/create_a_pipeline.md), you notice that Kedro generates a
+few boilerplate files: `nodes.py`, `pipeline.py`, `pipeline_registry.py`...
+
+While those may be sufficient for a small project, they quickly become large, hard to
+read and collaborate on as your codebase grows.
+Those files also sometimes make new users think that Kedro requires code
+to be located only in those starter files, which is not true.
+
+This section elaborates what are the Kedro requirements in terms of organising code
+in files and modules.
+It also provides examples of common scenarios such as sharing utilities between
+pipelines and using Kedro in a monorepo setup.
+
+## Where does Kedro look for code to be located
+
+The only technical constraint for arranging code in the project is that `pipeline_registry.py`
+file must be located in `<your_project>/src/<your_project>` directory, which is where
+it is created by default.
+
+This file must have a `register_pipelines()` function that returns a `tp.Dict[str, Pipeline]`
+mapping from pipeline name to corresponding `Pipeline` object.
+
+Other than that, **Kedro does not impose any constraints on where you should keep files with
+`Pipeline`s, `Node`s, or functions wrapped by `node`**.
+
+```{note}
+You actually can make Kedro look for pipeline registry in a different place by modifying the
+`__main__.py` file of your project, but such advanced customisations are not in scope for this section.
+```
+
+This being the only constraint means that you can, for example:
+* Add `utils.py` file to a pipeline folder and import utilities used by multiple
```

Comment:
@noklam how would you then call a `.py` file that is not `nodes.py` or `pipeline.py` for the purpose of this example? I was thinking of `dataframe_utils.py`, but didn't like it because it adds to the impression that Kedro is only useful in data processing projects, which isn't true.

Reply:
There is many thing that I don't like about `utils` (just me, not a team consensus). The purpose is ill-defined and often become the place that people dump code to without thinking. 

Even if we go with utils. I will strip the `_utils` suffix, it feels redundant to have `pandas_utils.py` under `utils.py`. Then in the code I will probably do ` from <pacakge> import utils`. When I need to use it, I will use `utils.pandas.func` to make it clear that this is a util namespace but not `pandas`.

`visualitization_utils.py` could just be `visualisation` module itself.
(all above are subjective)

I think it will be great to first introduce the principle of share module, what are the factors to consider. Then you can show this example. https://kedro-org.slack.com/archives/C03RKP2LW64/p1716912123397259
@datajoely do you have thought about this?

Reply:
- Comments re: naming utils modules and importing them - addressed ✔️
- Re: introducing the principles, the hyperlink doesn't work for me, leads just to the `questions` channel.